src/storage/src/object/disk.rs

```diff
@@ -0,0 +1,495 @@
+        let path = self.new_file_path(path)?;
+        let hash = {
+            let mut hasher = DefaultHasher::default();
+            path.hash(&mut hasher);
+            hasher.finish()
+        };
+        let entry = self
+            .opened_read_file_cache
+            .lookup_with_request_dedup(hash, path.clone(), || async {
+                let file = utils::open_file(&path, true, false, false)
+                    .await
+                    .map_err(HummockError::object_io_error)?
+                    .into_std()
+                    .await;
+                Ok((file, 1))
+            })
+            .await
+            .map_err(|e| {
+                ObjectError::disk(
+                    "".to_string(),
+                    std::io::Error::new(
+                        ErrorKind::Other,
+                        format!("Failed to open file {:?}. Err{:?}", path.to_str(), e),
+                    ),
+                )
+            })?;
+        Ok(Arc::new(entry))
+    }
+}
+
+impl Drop for LocalDiskObjectStore {
+    fn drop(&mut self) {
+        if let Some(sender) = self.compactor_shutdown_sender.lock().take() {
+            let _ = sender.send(());
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl ObjectStore for LocalDiskObjectStore {
+    async fn upload(&self, path: &str, obj: Bytes) -> ObjectResult<()> {
+        let mut file =
+            utils::open_file(self.new_file_path(path)?.as_path(), false, true, true).await?;
+        file.write_all(&obj)
+            .await
+            .map_err(|e| ObjectError::disk(format!("failed to write {}", path), e))?;
+        file.flush()
+            .await
+            .map_err(|e| ObjectError::disk(format!("failed to flush {}", path), e))?;
+        Ok(())
+    }
+
+    async fn read(&self, path: &str, block_loc: Option<BlockLocation>) -> ObjectResult<Bytes> {
+        match block_loc {
+            Some(block_loc) => Ok(self.readv(path, vec![block_loc]).await?.pop().unwrap()),
+            None => {
+                let file_holder = self.get_read_file(path).await?;
+                let metadata = utils::get_metadata(file_holder.clone()).await?;
+                let path_owned = path.to_owned();
+                utils::asyncify(move || {
+                    let mut buf = vec![0; metadata.len() as usize];
+                    file_holder
+                        .value()
+                        .read_exact_at(&mut buf, 0)
+                        .map_err(|e| {
+                            ObjectError::disk(
+                                format!("failed to read the whole file {}", path_owned),
+                                e,
+                            )
+                        })?;
+                    Ok(Bytes::from(buf))
+                })
+                .await
+            }
+        }
+    }
+
+    async fn readv(&self, path: &str, block_locs: Vec<BlockLocation>) -> ObjectResult<Vec<Bytes>> {
```

Comment:
`block_locs: impl AsRef<[BlockLocation]>`, so that users can either pass a reference `&[BlockLocation { xxx }]` or a vector.